File: Ternary logic/ternarity1.py
#!/usr/bin/env python3

class T:
    
    def __new__(self, arg=None):
        return self.T(arg)
    
    @staticmethod
    def T(arg):
        if arg is True: return True
        if arg is False: return False
        if arg is None: return None
        if arg == 0: return None
        if arg == 1: return True
        if arg == -1: return False
        if arg == 'TRUE': return True
        if arg == 'FALSE': return False
        if arg == 'NONE': return None
        if arg == 'NULL': return None
        if arg == 'UNDEFINED': return None
        if arg == 'True': return True
        if arg == 'False': return False
        if arg == 'None': return None
        if arg == 'Null': return None
        if arg == 'Undefined': return None
        if arg == 'true': return True
        if arg == 'false': return False
        if arg == 'none': return None
        if arg == 'null': return None
        if arg == 'undefined': return None
        if arg == 'N': return None
        if arg == 'n': return None
        if arg == 'T': return True
        if arg == 't': return True
        if arg == 'F': return False
        if arg == 'f': return False
        if arg == '0': return None
        if arg == '1': return True
        if arg == '-1': return False
        if isinstance(arg, (int, float)) and arg > 0: return True
        if isinstance(arg, (int, float)) and arg < 0: return False
        if arg == 'U': return None
        if arg == 'u': return None
        if arg == 'A': return True
        if arg == 'a': return True
        if arg == 'i': return False
        if arg == 'h': return False
        return None
    @staticmethod
    def T3(arg):
        return T.T(arg)
    
    @staticmethod
    def BIN(arg):
        return {
            None: False,
            True: True,
            False: False,
        }.get(T.T(arg))
    @staticmethod
    def T2(arg):
        return T.BIN(arg)
    
    @staticmethod
    def BIN1(arg):
        return {
            None: None,
            True: True,
            False: None,
        }.get(T.T(arg))
    
    @staticmethod
    def BIN2(arg):
        return {
            None: True,
            True: True,
            False: False,
        }.get(T.T(arg))
    
    @staticmethod
    def BIN3(arg):
        return {
            None: True,
            True: True,
            False: None,
        }.get(T.T(arg))
    
    @staticmethod
    def BIN4(arg):
        return {
            None: False,
            True: True,
            False: True,
        }.get(T.T(arg))
    
    @staticmethod
    def BIN5(arg):
        return {
            None: None,
            True: True,
            False: True,
        }.get(T.T(arg))
    
    @staticmethod
    def NOT(arg):
        return {
            None: None,
            True: False,
            False: True,
        }.get(T.T(arg))
    
    @staticmethod
    def NOT1(arg):
        return {
            None: False,
            True: True,
            False: None,
        }.get(T.T(arg))
    
    @staticmethod
    def NOT2(arg):
        return {
            None: True,
            True: None,
            False: False,
        }.get(T.T(arg))
    
    @staticmethod
    def LEFT(arg):
        return {
            None: False,
            True: None,
            False: True,
        }.get(T.T(arg))
    
    @staticmethod
    def RIGHT(arg):
        return {
            None: True,
            True: False,
            False: None,
        }.get(T.T(arg))
    
    @staticmethod
    def AND(arg1, arg2):
        return {
            (False, False): False,
            (False, None): False,
            (False, True): False,
            (None, False): False,
            (None, None): None,
            (None, True): None,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    @staticmethod
    def MIN(arg1, arg2):
        return T.AND(arg1, arg2)
    
    @staticmethod
    def OR(arg1, arg2):
        return {
            (False, False): False,
            (False, None): None,
            (False, True): True,
            (None, False): None,
            (None, None): None,
            (None, True): True,
            (True, False): True,
            (True, None): True,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    @staticmethod
    def MAX(arg1, arg2):
        return T.OR(arg1, arg2)
    
    @staticmethod
    def NAND(arg1, arg2):
        return {
            (False, False): True,
            (False, None): True,
            (False, True): True,
            (None, False): True,
            (None, None): None,
            (None, True): None,
            (True, False): True,
            (True, None): None,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def NOR(arg1, arg2):
        return {
            (False, False): True,
            (False, None): None,
            (False, True): False,
            (None, False): None,
            (None, None): None,
            (None, True): False,
            (True, False): False,
            (True, None): False,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def XAND(arg1, arg2):
        return {
            (False, False): True,
            (False, None): None,
            (False, True): False,
            (None, False): None,
            (None, None): None,
            (None, True): None,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    @staticmethod
    def EQV(arg1, arg2):
        return T.XAND(arg1, arg2)
    
    @staticmethod
    def XOR(arg1, arg2):
        return {
            (False, False): False,
            (False, None): None,
            (False, True): True,
            (None, False): None,
            (None, None): None,
            (None, True): None,
            (True, False): True,
            (True, None): None,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def XMIN(arg1, arg2):
        return {
            (False, False): True,
            (False, None): False,
            (False, True): False,
            (None, False): False,
            (None, None): True,
            (None, True): None,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def XMAX(arg1, arg2):
        return {
            (False, False): False,
            (False, None): None,
            (False, True): True,
            (None, False): None,
            (None, None): False,
            (None, True): True,
            (True, False): True,
            (True, None): True,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def WEBB(arg1, arg2):
        return {
            (False, False): None,
            (False, None): True,
            (False, True): False,
            (None, False): True,
            (None, None): True,
            (None, True): False,
            (True, False): False,
            (True, None): False,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def IMP(arg1, arg2):
        return {
            (False, False): True,
            (False, None): True,
            (False, True): True,
            (None, False): None,
            (None, None): None,
            (None, True): True,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def IMP1(arg1, arg2):
        return {
            (False, False): True,
            (False, None): True,
            (False, True): True,
            (None, False): None,
            (None, None): True,
            (None, True): True,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def IMP2(arg1, arg2):
        return {
            (False, False): True,
            (False, None): True,
            (False, True): True,
            (None, False): False,
            (None, None): True,
            (None, True): True,
            (True, False): False,
            (True, None): None,
            (True, True): True,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def MEAN(arg1, arg2):
        return {
            (False, False): False,
            (False, None): None,
            (False, True): False,
            (None, False): None,
            (None, None): True,
            (None, True): None,
            (True, False): False,
            (True, None): None,
            (True, True): False,
        }.get((T.T(arg1), T.T(arg2)))
    
    @staticmethod
    def MAGN(arg1, arg2):
        return {
            (False, False): None,
            (False, None): False,
            (False, True): False,
            (None, False): True,
            (None, None): None,
            (None, True): False,
            (True, False): True,
            (True, None): True,
            (True, True): None,
        }.get((T.T(arg1), T.T(arg2)))


#def ternarity(arg):
    #if arg is None:
        #print('None')
    #elif arg:
        #print('True')
    #else:
        #print('False')

#ternarity( T.T('-1') )      # False
#ternarity( T.T('T') )       # True
#ternarity( T.T(+123.45) )   # True
#ternarity( T.T('Text') )    # None
#ternarity( T.XAND( T.NOT(True), T.BIN(None) ) ) # True

#print(T(True))

File: Ternary logic/ternarity1.js

class T {

    static T(arg) {
        if (arg === true) return true;
        if (arg === false) return false;
        if (arg === undefined) return undefined;
        if (arg > 0) return true;
        if (arg < 0) return false;
        if (arg === 0) return undefined;
        if (arg === 'TRUE') return true;
        if (arg === 'FALSE') return false;
        if (arg === 'UNDEFINED') return undefined;
        if (arg === 'NULL') return undefined;
        if (arg === 'NONE') return undefined;
        if (arg === 'True') return true;
        if (arg === 'False') return false;
        if (arg === 'Undefined') return undefined;
        if (arg === 'Null') return undefined;
        if (arg === 'None') return undefined;
        if (arg === 'true') return true;
        if (arg === 'false') return false;
        if (arg === 'undefined') return undefined;
        if (arg === 'null') return undefined;
        if (arg === 'none') return undefined;
        if (arg === 'N') return undefined;
        if (arg === 'n') return undefined;
        if (arg === 'T') return true;
        if (arg === 't') return true;
        if (arg === 'F') return false;
        if (arg === 'f') return false;
        if (arg === 'U') return undefined;
        if (arg === 'u') return undefined;
        if (arg === '0') return undefined;
        if (arg === '1') return true;
        if (arg === '-1') return false;
        if (arg === 'A') return true;
        if (arg === 'a') return false;
        if (arg === 'i') return false;
        if (arg === 'h') return false;
        return undefined;
    }
    static T3(arg) {
        return this.T(arg);
    }
    
    static BIN(arg) {
        switch (this.T(arg)) {
            case undefined: return false;
            case true:      return true;
            case false:     return false;
        }
    }
    static T2(arg) {
        return this.BIN(arg);
    }
    
    static BIN1(arg) {
        switch (this.T(arg)) {
            case undefined: return undefined;
            case true:      return true;
            case false:     return undefined;
        }
    }
    
    static BIN2(arg) {
        switch (this.T(arg)) {
            case undefined: return true;
            case true:      return true;
            case false:     return false;
        }
    }
    
    static BIN3(arg) {
        switch (this.T(arg)) {
            case undefined: return true;
            case true:      return true;
            case false:     return undefined;
        }
    }
    
    static BIN4(arg) {
        switch (this.T(arg)) {
            case undefined: return false;
            case true:      return true;
            case false:     return true;
        }
    }
    
    static BIN5(arg) {
        switch (this.T(arg)) {
            case undefined: return undefined;
            case true:      return true;
            case false:     return true;
        }
    }
    
    static NOT(arg) {
        switch (this.T(arg)) {
            case undefined: return undefined;
            case true:      return false;
            case false:     return true;
        }
    }
    
    static NOT1(arg) {
        switch (this.T(arg)) {
            case undefined: return false;
            case true:      return true;
            case false:     return undefined;
        }
    }
    
    static NOT2(arg) {
        switch (this.T(arg)) {
            case undefined: return true;
            case true:      return undefined;
            case false:     return false;
        }
    }
    
    static LEFT(arg) {
        switch (this.T(arg)) {
            case undefined: return false;
            case true:      return undefined;
            case false:     return true;
        }
    }
    
    static RIGHT(arg) {
        switch (this.T(arg)) {
            case undefined: return true;
            case true:      return false;
            case false:     return undefined;
        }
    }
    
    static AND(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return false;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    static MIN(arg) {
        return this.T(arg);
    }
    
    static OR(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return undefined;
                    case true:      return true;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    static MAX(arg) {
        return this.T(arg);
    }
    
    static NAND(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static NOR(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return undefined;
                    case true:      return false;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return false;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static XAND(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return undefined;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    static EQV(arg) {
        return this.T(arg);
    }
    
    static XOR(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return undefined;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static XMIN(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return false;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return true;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    
    static XMAX(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return false;
                    case true:      return true;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static WEBB(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return true;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return false;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return false;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static IMP(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return undefined;
                    case true:      return true;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    
    static IMP1(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return true;
                    case true:      return true;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    
    static IMP2(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return true;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return true;
                    case true:      return true;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return true;
                }
            default: return undefined;
        }
    }
    
    static MEAN(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return true;
                    case true:      return undefined;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return false;
                    case undefined: return undefined;
                    case true:      return false;
                }
            default: return undefined;
        }
    }
    
    static MAGN(arg1, arg2) {
        switch (this.T(arg1)) {
            case false:
                switch (this.T(arg2)) {
                    case false:     return undefined;
                    case undefined: return false;
                    case true:      return false;
                }
            case undefined:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return undefined;
                    case true:      return false;
                }
            case true:
                switch (this.T(arg2)) {
                    case false:     return true;
                    case undefined: return true;
                    case true:      return undefined;
                }
            default: return undefined;
        }
    }
    
}


// function ternarity(arg) {
//     switch (T.T(arg)) {
//         case undefined: 
//             console.log('Undefined');
//             break;
//         case true:
//             console.log('True');
//             break;
//         case false:
//             console.log('False');
//             break;
//     }
// }
// 
// ternarity( T.T('-1') )      // False
// ternarity( T.T('T') )       // True
// ternarity( T.T(+123.45) )   // True
// ternarity( T.T('Text') )    // Undefined
// ternarity( T.XAND( T.NOT(true), T.BIN(undefined) ) ) // True


